### DISCUSSION TITLE
[Proposal][component style-guide] Tips and best practices for components development in Superset

### DISCUSSION THREAD

**Discussion by geido** [apache/superset#19304] [Ideas]
The following is a proposal to provide guidance on how best to approach component development in Superset.

***********************

This documentation illustrates how we approach component development in Superset and provides examples to help you in writing new components or updating existing ones by following our community-approved standards.

This guide is intended primarily for reusable components. Whenever possible, all new components should be designed with reusability in mind.

## General Guidelines

- We use [Ant Design](https://ant.design/) as our component library. Do not build a new component if Ant Design provides one but rather instead extend or customize what the library provides
- Always style your component using Emotion and always prefer the theme variables whenever applicable. See: [Emotion Styling Guidelines and Best Practices](https://github.com/apache/superset/wiki/Emotion-Styling-Guidelines-and-Best-Practices)
- All components should be made to be reusable whenever possible
- All components should follow the structure and best practices as detailed below

### Directory and component structure

```
superset-frontend/src/components
   {ComponentName}/ 
      index.tsx
      {ComponentName}.test.tsx
      {ComponentName}.stories.tsx
```

**Components root directory:** Components that are meant to be re-used across different parts of the application should go in the `superset-frontend/src/components` directory. Components that are meant to be specific for a single part of the application should be located in the nearest directory where the component is used, for example, `superset-frontend/src/Explore/components`

**Exporting the component:** All components within the `superset-frontend/src/components` directory should be exported from `superset-frontend/src/components/index.ts` to facilitate their imports by other components

**Component directory name:** Use `PascalCase` for the component directory name

**Storybook:** Components should come with a storybook file whenever applicable, with the following naming convention `{ComponentName}.stories.tsx`. More details about Storybook below

**Unit and end-to-end tests:** All components should come with unit tests using Jest and React Testing Library. The file name should follow this naming convention `{ComponentName}.test.tsx.` Read the [Testing Guidelines and Best Practices ](https://github.com/apache/superset/wiki/Testing-Guidelines-and-Best-Practices) for more details

**Reference naming:** Use `PascalCase` for React components and `camelCase` for component instances

**BAD:** `import mainNav from './MainNav;`
**GOOD** `import MainNav from './MainNav;`

**BAD**: `const NavItem = <MainNav />;`
**GOOD**: `const navItem = <MainNav />;`

**Component naming:** Use the file name as the component name

**BAD**: `import MainNav from './MainNav/index';`
**GOOD**:`import MainNav from './MainNav;`

**Props naming:** Do not use DOM related props for different purposes

**BAD**: `<MainNav style="big" />`
**GOOD**: `<MainNav variant="big" />`

**Importing dependencies:** Only import what you need

**BAD**: `import * as React from "react";`
**GOOD**: `import React, { useState } from "react";`

**Default VS named exports:** As recommended by [Typescript](https://www.typescriptlang.org/docs/handbook/modules.html ), “If a module’s primary purpose is to house one specific export, then you should consider exporting it as a default export. This makes both importing and actually using the import a little easier”. If you’re exporting multiple objects, use named exports instead.

_As a default export_
`import MainNav from './MainNav';`

_As a named export_
`import { MainNav, SecondaryNav } from './Navbars;`

**ARIA roles:** Always make sure you are writing accessible components by using the official [ARIA roles](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques)

## Use TypeScript

All components should be written in TypeScript and their extensions should be `.ts` or `.tsx `

### type vs interface

Validate all props with the correct types. This replaces the need for a run-time validation as provided by the prop-types library.

`
type HeadingProps = {  
   param: string;
}
`

`
export default function Heading({ children }: HeadingProps) {
  return <h2>{children}</h2>
}
`

Use `type` for your component props and state. Use `interface` when you want to enable _declaration merging_.

### Define default values for non-required props

In order to improve the readability of your code and reduce assumptions, always add default values for non required props, when applicable, for example:

`
const applyDiscount = (price: number, discount = 0.05) => price * (1 - discount);
`

## Functional components and Hooks

We prefer functional components and the usage of hooks over class components.

### useState

Always explicitly declare the type unless the type can easily be assumed by the declaration.

`
const [customer, setCustomer] = useState<ICustomer | null>(null);
`

### useReducer

Always prefer `useReducer` over `useState` when your state has complex logics.

### useMemo and useCallback

Always memoize when your components take functions or complex objects as props to avoid unnecessary rerenders.

### Custom hooks

All custom hooks should be located in the directory `/src/hooks`. Before creating a new custom hook, make sure that is not available in the existing custom hooks.

## Storybook

Each component should come with its dedicated storybook file.

**One component per story:** Each storybook file should only contain one component unless substantially different variants are required

**Component variants:** If the component behavior is substantially different when certain props are used, it is best to separate the story into different types. See the `superset-frontend/src/components/Select/Select.stories.tsx` as an example.

**Isolated state:** The storybook should show how the component works in an isolated state and with as few dependencies as possible

**Use args:**  It should be possible to test the component with every variant of the available props. We recommend using [args](https://storybook.js.org/docs/react/writing-stories/args)

**Comment by geido** [comment]
Converting this into a discussion

**Comment by ktmud** [comment]
Thanks for putting this together!

Some immediate thoughts:
- For code style choices, one should always just follow Prettier and ESLint. If something is not banned by ESLint, then it's allowed. If something gives warning in ESLint, then please fix it.
- If something will be captured by ESLint, then we don't need to include it in this guide. One example would be discouraging using loop indexes as keys. Ideally we'd like this style to be short and sweet and that people can remember and learn something new after reading it.
- We should guide users to use existing hooks in `/src/hooks` and add new ones in this centralized place.
- I'd recommend changing "We prefer the usage of functional components" to "we prefer functional components over class components" just for clarity. 

----

Re: code readability, allow me to rewrite your code to demonstrate a couple of points:

```ts
const Heading = ({ param }: HeadingProps): JSX.Element => <div>{param1}</div>;

function applyDiscount(price: number, discount?: number = 0.05): number { return price * (1 - discount); }
```

can be rewritten to

```ts
export default function Heading({ children }: HeadingProps) {
  return <h2>{children}</h2>
}

const applyDiscount = (price: number, discount = 0.05) => price * (1 - discount);
```

1. To maximum readability and reduce footprint, return value types should be avoided unless it adds clarify.
2. Arrow functions are preferred for one-liners---ESLint will warn you about this.
3. Optional parameters with default values do not need `?` and type annotation. TypeScript is smart enough to infer `discount` is `number`.
4. I _personally_ prefer to write React component as named functions and utility functions with arrow functions as it makes React components looks more "special" and code is more concise this way when you `export default`.

**Reply by diegomedina248** [reply]
On point 4, I prefer the opposite, but I value consistency above all, so I think we should make a decision and enforce that style choice.

**Reply by ktmud** [reply]
I wonder if there's an ESLint or Prettier rule to take care of the auto-reformatting for us. If not, then let's not worry about it. The whole point of using these tools is so we don't argue about styles.

**Reply by diegomedina248** [reply]
I think [this](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/function-component-definition.md) could work

**Reply by geido** [reply]
I tend to prefer the opposite as well but I am curious of what other people think

**Reply by ktmud** [reply]
I'm ok with picking either as long as it can be autofixed by linter.

**Comment by eschutho** [comment]
@geido This is great. Should we make a distinction between reusable components and larger components that are made up of smaller components? For example, creating a Story would apply to a reusable component like a dropdown component but a Cypress test would be applicable for a larger component that makes up a larger portion of the page as in the SqlLab left toolbar component. I think focusing this document on one or the other could make it simpler.

**Reply by geido** [reply]
Thanks! I received this feedback before as well. I edited the guide to be specific for reusable components.

**Comment by rusackas** [comment]
Awesome stuff! The one nit I might want to change is the bit about Default Exports. I kind of prefer Named Exports. Even if you `import { SomeComponent as WhateverYouWantToCallIt } from './SomeFile'` it's very easy to search globally in your IDE for instances of `SomeComponent` whereas that's more difficult when if someone does `import YourThingWasRenamedHere from  './SomeFile'`.

**Reply by geido** [reply]
I like this suggestion actually but I think this falls under this:

> Component naming: Use the file name as the component name

It should have the same effect. If others have strong opinions about this we can definitely revisit it.
